formal verification: Why3 translation of mapping types

pirapira opened the pull request:
Solves #1025 .

After this PR, a state variable `mapping (address => uint256) balance;` is translated into

```
        type state = {
                mutable _balance: array uint256
        }
```

Since the Why3 `array` allows negative indices, the translation works for any `IntegerType`, including signed integers.

While doing that, I wanted an assertion over a shared pointer, so the `assertThrow` macro is also changed.

pirapira committed:
Explicitly convert the assert condition using bool()

This allows a shared pointer as a condition because bool(std::shared_ptr<T>) is defined.

pirapira committed:
Translate mapping types into Why3 arrays when keys are integers

Even when the keys are signed the translation is supposed to work
because Why3 arrays allow negative indices.

chriseth commented on libdevcore/Assertions.h:
```diff
@@ -73,7 +73,7 @@ inline bool assertEqualAux(A const& _a, B const& _b, char const* _aStr, char con
 /// Do NOT supply an exception object as the second parameter.
 #define assertThrow(_condition, _ExceptionType, _description) \
-	::dev::assertThrowAux<_ExceptionType>(_condition, _description, __LINE__, __FILE__, ETH_FUNC)
+	::dev::assertThrowAux<_ExceptionType>(bool(_condition), _description, __LINE__, __FILE__, ETH_FUNC)
```
We usually use `assert(!!x, "msg")` if the compiler complains. Is there a situation where an explicit `bool(...)` is more permissive than we would like?

pirapira replied:
`!!` also works for me.  So the question is what has `bool()` but not `!`... I don't know!  I follow your general custom, and change this into `!!`.

chriseth commented on libsolidity/formal/Why3Translator.cpp:
```diff
@@ -77,12 +77,26 @@ string Why3Translator::toFormalType(Type const& _type) const
+			string valueType = toFormalType(*mappingType->valueType());
+			if (!valueType.empty())
+				return "array " + valueType;
+		}
```
else report an error

pirapira replied:
I will add `else` in this PR here.  I will also add it in the above clauses in separate PRs.

pirapira replied:
I don't have an `ASTNode` here, but I want to use `Why3Translator::error(ASTNode const&, string)`.  So either I have to
1. add an `ASTNode` to this function; or
2. raise an exception here and add `ASTNode` information in a caller
I'll try 2. first.

chriseth commented on libsolidity/formal/Why3Translator.cpp:
```diff
@@ -77,12 +77,26 @@ string Why3Translator::toFormalType(Type const& _type) const
+		solAssert(mappingType->keyType(), "A mappingType misses a keyType.");
+		if (dynamic_cast<IntegerType const*>(&*mappingType->keyType()))
+		{
+			//@TODO Use the information from the key type and specify the length of the array as an invariant.
```
Do you want to specify the length as `2**256` or just leave it unspecified (but constant)?

pirapira replied:
If the `keyType` is `uint256`, then the length would be `2**256`, but when `keyType` is `address` the length would be `2**160`.  This would automatically add verification conditions that check balances are checked only for small numbers.

pirapira committed:
toFormalType reports errors by an exception

This allows error reporting without passing `ASTNode` to `toFormalType()`

pirapira committed:
Prepare for leaky exceptions

Now toFormalType() reports errors by exceptions, they will be
sometimes leaked to the wider context.  This commits adds a catch.

pirapira commented:
mappings to mappings do not work currently.  They will be dealt in #1053.

pirapira commented:
I've addressed the comments and the tests have passed.

chriseth commented on libsolidity/formal/Why3Translator.cpp:
```diff
@@ -434,8 +494,15 @@ bool Why3Translator::visit(TupleExpression const& _node)
+	catch (NoFormalType &err)
+	{
+		string const* typeNamePtr = boost::get_error_info<errinfo_noFormalTypeFrom>(err);
+		error(_unaryOperation, "Type not " + (typeNamePtr ? *typeNamePtr : "") + " supported in unary operation.");
```
I think the `not` is in the wrong place.

pirapira replied:
:confounded: I will fix this.

chriseth commented:
The change in error condition looks more verbose now, but should be more concise once we have better coverage of the types, so good change :-)

pirapira commented:
Addressed comments.
